### ISSUE TITLE
Cannot link to non-md files anymore

### ISSUE THREAD

**Issue by paperbenni** [vimwiki/vimwiki#1271]
I have some latex files stored and linked to in my wiki.
Before https://github.com/vimwiki/vimwiki/commit/d7ec12645a0460a7d200279c52915e6e080e9869 I could omit the `file:` part of a link to have vim open the file directly. 
After https://github.com/vimwiki/vimwiki/commit/d7ec12645a0460a7d200279c52915e6e080e9869  vimwiki adds a md extension when opening any file which results in a filename.tex.md being created alongside the original tex file. Using `file:` is not a solution since this can only be configured to open a second instance of vim. 
I would propose not adding the extension if `filename.otherextension` exists but `filename.otherextension.md` does not. That way people can link to md files with the same name as other files but retain the ability to directly open these files from their wiki. 

```
let g:vimwiki_markdown_link_ext=0
let wiki = {}
let wiki.path = '~/wiki/vimwiki'
let wiki.syntax = 'markdown'
let wiki.ext = '.md'
let wiki.automatic_nested_syntaxes = 1
let streamwiki = {}
let streamwiki.path = '~/streamwiki/vimwiki'
let streamwiki.automatic_nested_syntaxes = 1
let g:vimwiki_conceal_pre = 1
let g:vimwiki_list = [wiki, streamwiki] " wikilistmarker

```

```
Os: Linux
Vim: 800
Branch: dev
Revision: 584f073
Date: 2022-08-23 16:48:01 -0600
```

**Comment by brennen** [comment]
Sorry for the frustrating change in behavior with d7ec126.

I don't really think this is a bug _as such_ so much as an example of [Hyrum's Law](https://www.hyrumslaw.com/) in action. Making things more DWIM here by checking for the existence of a file would probably be defensible, but it feels kind of fuzzy in a way that makes me nervous.  More importantly, it would complicate `vimwiki#base#resolve_link()`, which is already a hard part of the code to change, and make it impossible to parse VimWiki pages without knowing the set of files on disk.

As a suggestion, here's what I did in my personal config to make `file:` links more useful:

```vim
    " override default vimwiki link handling - could be extended to add
    " custom link types, etc.  Right now all it does is use vim itself
    " to handle files, crudely.
    function! VimwikiLinkHandler(link)
      if a:link =~# '^file:'
        try
          " chop off the leading file: - see :h expr-[:] for syntax:
          execute ':split ' . a:link[5:]
          return 1
        catch
          echo "Failed opening file in vim."
        endtry
      endif
      return 0
    endfunction
```

**Comment by paperbenni** [comment]
The problem with that approach is that I also have non-md files in my wiki that are actual binaries, I don't want vim to open my pdf files. Maybe adding another prefix like `file:` for non-md plain text files would be an option. Off the top of my head I could think of `text, plain, txt` or  `vim` as name ideas

**Comment by brennen** [comment]
Again, I'd suggest handling that in a custom link handler. Here's my current full one as a working example of handling more than one prefix:

```vim
    " override default vimwiki link handling - could be extended to add
    " custom link types, etc.  Right now all it does is use vim itself
    " to handle files, crudely.
    function! VimwikiLinkHandler(link)
      if a:link =~# '^file:'
        try
          " chop off the leading file: - see :h expr-[:] for syntax:
          execute ':split ' . a:link[5:]
          return 1
        catch
          echo "Failed opening file in vim."
        endtry
      elseif a:link =~# '^workspace:'
        try
          " chop off the leading file: - see :h expr-[:] for syntax:
          execute ':split ' . '~/workspace/' . a:link[10:]
          return 1
        catch
          echo "Failed opening workspace location in vim."
        endtry
      " elseif a:link =~# '^https\?:'
      "    try
      "      execute ':terminal lynx ' . a:link
      "      return 1
      "    catch
      "      echo "Failed executing Lynx."
      "    endtry
      endif
      return 0
    endfunction
```

**Comment by FIGBERT** [comment]
Anecdotally, this change has been a big negative for my workflow. I have been using Vimwiki as a means to organize and link between a wide network of `UUID`-named binary, `.wiki`, and text files.

The binary workflow still works, with the `file:` prefix that opens it using macOS' `open`. But for other files, where I (for example) link to a series of `csv`s, it's completely broken. I used to be able to open and edit arbitrary text formats from wiki files, which now requires bespoke configuration in my `init.lua` and deactivates the convenient Vimwiki back-and-forth between files with `delete` and `enter`.

I think this merits addressing in some form, and I would certainly be happy to help in any way.

**Comment by brennen** [comment]
Yeah, that's fair.  This clearly has users — I'll reopen and give some thought to a least-disruptive sort of solution.

I'm loathe to expand VimWiki as a language too much at this stage of things, but it's possible that we just need to define an explicit out of the box link type for "open this in the editor". It certainly comes up enough.  I'm not sure about the part where back/forward still works.  "Open this non-wiki thing in the editor but treat it as part of the wiki" is clearly desirable in some sense, but at the moment I suspect it works more by happenstance than by intent.

Some exploration to be done.

**Comment by tinmarino** [comment]
Hello @brennen,

I guess I spammed you a little this week for other tickets, sorry for that.
I would suggest to add a flag to trigger the new behavior. I also prefer the old behavior which was:

1. Adding a .md extension to the target only if there were no extension (i.e. not for my_script.py)
2. The LinkHandler do not search for a file with an added .md systematically

With the new behaviour, even if I remove by hand the .md suffix automatically added, I still go to the my_script.py.md file. So impossible to go automatically to my_script.py. 

I agree with point that the fall back makes the converted version uncertain in [your comment](https://github.com/vimwiki/vimwiki/issues/1271#issuecomment-1272727996). (good point I would not think about it).

Which is why I like the boolean flag.

1. You always use .md files (never .py or .bash, .cpp ... or in those rare cases, you use the file: prefix )
2. I never use dot in my file names except for extensions so if there is an extension (like .py) it is the good one

Basically old vs new, with the old as default.

If you agree, feel free to create the flag variable, or assign me the ticket, I can code it too looking at those 2 changes.

And finally, thank you very much for your consistence in maintaining vimwiki where other maintainers, like me are absent for long periods (years), you keep having an eye and that helps a lot.

**Comment by brennen** [comment]
Just a brief note that I'm thinking through some options here. Mostly by scribbling in a notebook about pros and cons, at the moment.

Further updates as events warrant.

**Comment by brennen** [comment]
I bashed out a bit of a decision doc here.  **Consider this a draft; it'll probably live somewhere else eventually.** I'm calling it a night, but wanted to put something here in the meanwhile.

# context

<https://github.com/vimwiki/vimwiki/issues/950>

<https://github.com/vimwiki/vimwiki/commit/d7ec12645a0460a7d200279c52915e6e080e9869>

Commit d7ec126 applies the wiki extension to files containing a `.`, thus allowing page names to contain a literal period. This is correct behavior, insofar as pages like `Franklin D. Roosevelt` should be able to exist and get mapped to e.g. `Franklin D. Roosevelt.wiki`.

Unfortunately, this broke behavior that a number of users rely on:

<https://github.com/vimwiki/vimwiki/issues/1271>

> I have some latex files stored and linked to in my wiki.
>
> Before d7ec126 I could omit the file: part of a link to have vim open the file directly.
>
> After d7ec126 vimwiki adds a `md` extension when opening any file which results in a `filename.tex.md` being created alongside the original tex file. Using `file:` is not a solution since this can only be configured to open a second instance of vim.

It makes sense that this is useful.

Sidebar: `file:` *can* be configured to do just about anything with a custom handler, but that doesn\'t mean it\'s a very good user experience.

# possible solutions

## revert to original behavior

As a default, I personally think this is a non-starter. Not being able to use `.` in page names is broken. I also suspect other things break if you treat those files as wiki pages.

We could make it configurable, though.

**Pro:**

-   People can go back to their expected behavior by setting one config variable.

**Con:**

-   Adds an extra branch to link handling code.
-   I haven\'t thought through the implications of treating it as fully supported behavior. What else will people expect to work right?

## check if a file exists on disk when following link

The idea is that if a link is `[[foo.txt]]` and `foo.txt` already exists, we should edit the existing file instead of creating a `foo.txt.md`.

**Pro:**

-   Might be the closest to expectations of people who rely on the old behavior.
-   Doesn\'t expand the language itself.

**Con:**

-   Complicates link handling code.
-   Parsing a page requires knowing about external files.
-   Might be surprising / confusing. You can write `foo.txt` and wind up with two different outcomes depending on whether `foo.txt` already exists.
    -   It\'s very Do What I Mean, which isn\'t necessarily bad. But if I write a link to `foo.txt` and wind up at `foo.txt.md`, it\'s not obvious that\'s what I meant.

## add a link scheme that means \"open in vim\"

Something like `[[edit:foo.txt]]`.

I vaguely think there\'s prior art for this in some older browsers? I could swear Lynx had something like it, although I couldn\'t find anything under [wp:List of URI schemes](https://en.wikipedia.org/wiki/List_of_URI_schemes) or the [list of Lynx URL schemes](https://lynx.invisible-island.net/lynx_help/lynx_url_support.html).  I may just be thinking of `lynxexec:`.

**Pro:**

-   People want this anyway, and a `:VimwikiLinkHandler` for using `file:` or some custom prefix to handle it is a common tweak to configurations.
-   It doesn\'t need a file existence check.

**Con:**

-   Expands the language.
-   How is this converted to HTML?
    -   What will Markdown parsers do with it?

# open questions

-   What is the actual current behavior with Markdown?
    -   Does it differ meaningfully from vimwiki syntax?
    -   If so, why?

**Comment by redongh** [comment]
I did a little testing in regard to your open questions in the hope that it might be of help.

Assume this [index.mdw](https://gist.github.com/redongh/ea111445631f4ee65e3c635c4aafc324#file-index-mdw) of my vimwiki (see content for specifics of the setup) making use of custom `VimwikiLinkHandler` for `file:` and `:raw` links.

Within vimwiki itself, both variants work as expected (= are being openend in a split within vim).

Converting to _HTML_ through the `custom_wiki2html` yields a index.html where
* original file-extensions are replaced by `.html` 
* none of the generated links to external files work:
  * `<li>a file link to <a href="file://$HOME/testsubdir/external.html">external-txtfile</a></li>` 
  * `<li>a raw link to <a href="raw:testsubdir/external.html">external-txtfile</a></li>`

Wouldn't it be desireable if the `:file` link when converted to _HTML_ would look like this:  
`<li>a file link to <a href="file:///home/user/testsubdir/external.txt">external-txtfile</a></li>`  ?

That doesn't seem too far from what is being generated now, I'm just not sure where this processing should ideally take place...

Regarding the possible solutions outlined by @brennen above, I get quite an itchy feeling when it comes to checking for the existence of files with different extensions as I'm afraid that this might cause lagging for pages with larger link collections.

In my current setup, the vimwiki-page with the most links to non-vimwiki files has currently over 200 links inside (i was quite surprised about the amount myself), so it is something I would say is being easily possible to be encountered.

Therefore I (as a clear vimwiki newbie) would lean towards the other two options but then in turn it would have to be determined how the conversion of links in _HTML_ should be handled or what the options are here. Similar thing with _PDF_-targets, I usually prefer converting links to footnotes so the information is not lost when those documents are being printed.

**Comment by brennen** [comment]
Thanks @redongh.

I updated my above decision doc comment with a few more notes and a new option, **add configuration for a set of filetypes to open in vim**.

On thinking through this all a bit more, that's the one I'll probably try hacking together first.  It would require changes to existing documents to get the desired behavior, but aside from that it seems like it avoids most of the downsides of the other options. It would also cover a lot of questions I've seen asked on IRC, and let people simplify their setups. (At least it would let me simplify mine, and I suspect other people are doing the same sort of thing.)

Thoughts welcome, of course.

**Comment by octaskin** [comment]
Just my 2 cents (maybe not even that much), on the issue.

Letting the link work as [Visible Name](relative-link-to-file.pdf) without the file: tag is rendered by Github as well. Which might be a small bonus to ppl who are sharing their wiki with other users without vimwiki. 

Sorry if my lame usecase making noise on the constructive discussion.

**Comment by brennen** [comment]
> Letting the link work as [Visible Name](relative-link-to-file.pdf) without the file: tag is rendered by Github as well, as here->[Visible Name](https://github.com/vimwiki/vimwiki/issues/relative-link-to-file.pdf). Which might be a small bonus to ppl who are sharing their wiki with others who might not even know about vimwiki.

Ugh. Yeah, like, this is clearly a correct thing when considered in the light of _Markdown_, the (fuzzy) standards/expectations for which diverge in all these subtle ways from how VimWiki is designed, how the better-supported native syntax works, and how VimWiki works as a thing primarily experienced _in the editor_ as vs. _when rendered to HTML_, which is an important use case for some people and almost irrelevant to how others interact with it.

For approximately the thousandth time, I wish this project either didn't attempt to support more than one markup language, or had been designed from the ground up to support more than one markup language cleanly, which would imply some very different patterns I think.

As it is, well, here we are I guess.  There are some things I just don't know if it's possible to easily reconcile. At least not without a huge amount of tricky special-casing.

Anyway, I'll keep this in mind when next looking through the code again. Maybe it's not as hard to square the circle as I think it is. Or maybe the right thing to do would be to rip out the current link resolution code and painstakingly rebuild it from the ground up to make this all easier to reason about.

**Comment by brennen** [comment]
(Hope I don't sound too testy in the above. This is just one of those places with a ton of different conflicting needs and expectations build into the problem.)

**Comment by FIGBERT** [comment]
@brennen: I think your point regarding the varying usage of VimWiki is prescient. I use VimWiki extensively, but entirely from NeoVim—I have never once even *attempted* to convert my wiki to HTML.

I think this use case makes my initial idea of a behavior summarized as "open this non-wiki thing in the editor but treat it as part of the wiki" a little more understandable. From my perspective, these non-`.wiki` files are not "non-wiki thing"s, they are just parts of the wiki that happen to be in different formats (like a `.csv`).  Anything that I link to and interact with from the editor (text or binary) is part of the wiki! For someone who renders their wiki to HTML, I can understand that there is more of a distinction.

It is possible that opening non-`.wiki` files as part of the wiki is out of the scope of the project. I personally think it has value, but I would understand if I have just been exploiting an edge-case.

---

As someone who has used `.wiki` files from the beginning, I am less familiar with the complexities of the multi-language markup support of the project.

---

With regard to the design doc above, I will note a few things.

I briefly explored [Org Mode](https://orgmode.org/) through [Doom Emacs](https://github.com/doomemacs/doomemacs) when this change initially dropped, and it appears to have opted for the additional-link-scheme route.

I am not a fan of the check-if-the-file-exists method for all the reasons noted above, nor the config-toggle-for-file-links as I would run in to the immediate issue of how to mix `.pdf`s and `.csv`s with the `file:` link scheme.

The config for file types is a fairly palatable solution.

---

I appreciate the care with which you are approaching this issue, and offer both any assistance I can provide, and my condolences for the fact that no matter what happens it will likely break *someone's* workflow (hopefully minimally).

May we walk the shrouded path with wisdom.

**Comment by fullstopslash** [comment]
I'd like to just chime in I found this issue after spending hours trying to figure out why vimwiki insist's on adding a .md to a file even when I have ``` vim.g.vimwiki_markdown_link_ext = 0``` set.   I'm using vimwiki with markdown,  LaTeX files , some encrypted notes {.gpg} , and links to various other filetypes that are all editable in neovim, and it's been really confusing as to why it insists on appending a .md when I set the option explicitly to not do so. I'd love to throw money at someone for solving this issue, but I don't see a donate button on this git repo!

**Comment by tinmarino** [comment]
Hi guys, thank Brennen for [the proposed options, very nice comment](https://github.com/vimwiki/vimwiki/issues/1271#issuecomment-1482207680).

I made a "fix" for this ticket at PR #1418. Can some mantainer (@brennen) have a look please?

There is no good option so I chose the easiest one, just add a boolean configuration variable (called **open_link_add_ext**) and let user decide.

I like the possibility to visit .sh .py .txt files.
We can see in the comments of this ticket that some people, like me, would prefer "not adding" and in 3 other tickets referenced in d7ec12645a04, we can see that other people prefer "adding" the extension.

**Comment by tinmarino** [comment]
Hi @brennen (or other maintainer),

I added very small switch so that I can configure the .wiki extension automatically added at the end.

It is a small change, I did not create test, not release. The code is not very pretty but spare enough to read.

If that is OK for you too, you can accept the PR.

**Comment by brennen** [comment]
> I made a "fix" for this ticket at PR https://github.com/vimwiki/vimwiki/pull/1418. Can some mantainer (@brennen) have a look please?

Thanks for picking this up - I'll allocate some time this weekend to review and get some things merged here.
